singularity-eos/base/table_bounds.hpp

```diff
@@ -0,0 +1,285 @@
+
+    grid = gridFromIntervals_(ThreeGrids(), global_min, global_max, anchor_point, mid_min,
+                              mid_max, ppd_fine, ppd_factor_lo, ppd_factor_hi);
+  }
+
+  inline Real log2lin(const Real xl) const {
+    // JMM: Need to guard this with the linear bounds passed in. The
+    // reason is that our fast math routines, while completely
+    // invertible at the level of machine epsilon, do introduce error
+    // at machine epsilon, which can bring us out of the interpolation
+    // range of eospac.
+    return std::min(linmax_,
+                    std::max(linmin_, singularity::FastMath::pow10(xl) - offset));
+  }
+  inline Real i2lin(const int i) const { return log2lin(grid.x(i)); }
+
+  friend std::ostream &operator<<(std::ostream &os, const Bounds &b) {
+    os << "Bounds: [" << b.grid.min() << ", " << b.grid.max() << "]"
+       << " + " << b.offset << "\n"
+       << "\tN = " << b.grid.nPoints() << "\n";
+    for (int ig = 0; ig < b.grid.nGrids(); ++ig) {
+      os << "\t[ig,dx] = [" << ig << ", " << b.grid.dx(ig) << "]"
+         << "\n";
+    }
+    return os;
+  }
+
+  // This uses real logs
+  template <typename T>
+  static int getNumPointsFromPPD(Real min, Real max, const T ppd) {
+    constexpr Real epsilon = std::numeric_limits<float>::epsilon();
+    const Real min_offset = 10 * std::abs(epsilon);
+    // 1.1 so that the y-intercept isn't identically zero
+    // when min < 0.
+    // min_offset to handle the case where min=0
+    Real offset = 0;
+    if (min <= 0) offset = 1.1 * std::abs(min) + min_offset;
+    min += offset;
+    max += offset;
+
+    Real lmin = std::log10(min);
+    Real lmax = std::log10(max);
+    int N = getNumPointsFromDensity(lmin, lmax, ppd);
+    return N;
+  }
+  template <typename T>
+  static int getNumPointsFromDensity(const Real min, const Real max, const T density) {
+    Real delta = max - min;
+    int N = std::max(2, static_cast<int>(std::ceil(density * delta)));
+    return N;
+  }
+
+ private:
+  Real toLog_(Real val, const Real offset) {
+    val += offset;
+    return singularity::FastMath::log10(std::abs(val));
+  }
+
+  void convertBoundsToLog_(Real &min, Real &max, const Real shrinkRange = 0) {
+    // Log scales can't handle negative numbers or exactly zero. To
+    // deal with that, we offset.
+    constexpr Real epsilon = std::numeric_limits<float>::epsilon();
+    const Real min_offset = 10 * std::abs(epsilon);
+    // 1.1 so that the y-intercept isn't identically zero
+    // when min < 0.
+    // min_offset to handle the case where min=0
+    if (min <= 0) offset = 1.1 * std::abs(min) + min_offset;
+
+    min = toLog_(min, offset);
+    max = toLog_(max, offset);
+
+    Real delta = max - min;
+    min += 0.5 * shrinkRange * delta;
+    max -= 0.5 * shrinkRange * delta;
+  }
+
+  static void adjustForAnchor_(const Real min, Real &max, int &N,
+                               const Real anchor_point) {
+    if (min < anchor_point && anchor_point < max) {
```

Comment:
What are the implications if this condition isn't met, would it make sense to error?

Reply:
If the condition isn't met, it will just not adjust the number of points or the max. The table will still be valid, but there won't be a grid point exactly at the anchor point. I'd like `sesame2spiner` to basically always run... so I'd rather this not error out. But I'll add a warning.